### PR TITLE
[libcxx] Fix 'transform_error.mandates.verify.cpp' test on msvc

### PR DESCRIPTION
**Pull request by zeroomega**
PR #102851 marks reference types in union as error on msvc by changing the clang, which makes 'transform_error.mandates.verify.cpp' no longer failing on msvc from ToT. However, all libcxx buildbots do not build clang from source, therefore, this test will still fail on these bots, which is incorrect. This patch changed the expected error message of this test so it can pass with both release branch clang and ToT clang.

### CODE REVIEW

File: libcxx/test/libcxx/utilities/expected/expected.void/transform_error.mandates.verify.cpp
Comment (ilovepi) [inline libcxx/test/libcxx/utilities/expected/expected.void/transform_error.mandates.verify.cpp:18]: We should probably split the msvc bit out into its own `UNSUPPORTED:` group w/ this comment and a tracking issue on github, since once libcxx CI builds with a new enough clang, this could go back to being supported. Same for the other test.

**Review by ilovepi — APPROVED** [review]
LGTM, w/ minor comment. Let's also get feedback from one of the libc++ maintainers before landing.

File: libcxx/test/libcxx/utilities/expected/expected.void/transform_error.mandates.verify.cpp
Comment (zeroomega) [reply libcxx/test/libcxx/utilities/expected/expected.void/transform_error.mandates.verify.cpp]: I have moved it into a new unsupported group.
For filing an upstream tracking issue, what content should I put into it? We can file a bug to ask rolling the clang in libcxx bots, but libcxx bots uses the latest LLVM release so they will be rolled once llvm has a new release. It would be great that all libcxx bots build clang from ToT but I am not sure if that fits libcxx maintainer's best interests.

File: libcxx/test/libcxx/utilities/expected/expected.expected/transform_error.mandates.verify.cpp
Comment (philnik777) [inline libcxx/test/libcxx/utilities/expected/expected.expected/transform_error.mandates.verify.cpp:16]: I think what we really want to do is add the diagnostics with a `expected-error 0-1` for the MSVC diagnostics and a comment that we can remove things in LLVM 22(?). As-is this will just result in the test being `UNSUPPORTED` indefinitely on MSVC.

Comment (zeroomega) [reply libcxx/test/libcxx/utilities/expected/expected.expected/transform_error.mandates.verify.cpp:16]: Could you elaborate it a bit please? I am not very familiar syntax of `expected-error` and the '0-1' feature wasn't present in lit's document. My concern is that modify the test to use `expected-error 0-1` would require modifying every single test case. And ideally, adding a comment at each line that is modified. It is not a clean solution 
IMO, the correct way to go forward is t overhaul libcxx test infrastructure to use ToT clang instead of release clang. It can solve a lot of inconsistency issues we have seen so far. But we all not that cannot be done in one day. Disabling this test and waiting for a clang roll on libcxx side is way going forward.

Comment (philnik777) [reply libcxx/test/libcxx/utilities/expected/expected.expected/transform_error.mandates.verify.cpp:16]: If the diagnostic is always the same you can also make it `0-8`. It simply says we expect `0-1` errors. If there are different errors you can also just add a single `TODO` at the top of the file. Also, we should absolutely test with release versions of Clang, since we support them. We test _also_ with trunk.

Comment (zeroomega) [reply libcxx/test/libcxx/utilities/expected/expected.expected/transform_error.mandates.verify.cpp:16]: > If the diagnostic is always the same you can also make it `0-8`. It simply says we expect `0-1` errors. If there are different errors you can also just add a single `TODO` at the top of the file. Also, we should absolutely test with release versions of Clang, since we support them. We test _also_ with trunk.

Can I make the expect `0-1` errors only on msvc, what would the syntax looks like? For other platforms, test should through an error otherwise it should fail, we should guarantee that.

>  Also, we should absolutely test with release versions of Clang, since we support them. We test _also_ with trunk.
Not with Windows. As far as I can tell all windows bots for libcxx use release branch of clang. If they use ToT clang the test infra should have caught this issue.

Comment (philnik777) [reply libcxx/test/libcxx/utilities/expected/expected.expected/transform_error.mandates.verify.cpp:16]: I don't think it's necessary to restrict it to a specific platform, but if you want to you should be able to just add `#ifdef _WIN32` around it. We're not testing clang here after all. The main thing we want to test here are the `static_assert`s.

Comment (zeroomega) [reply libcxx/test/libcxx/utilities/expected/expected.expected/transform_error.mandates.verify.cpp:16]: I have modified this change using `expected-error-re 0-1` and `expected-warning-re 0-1`. And added TODO at the top. Please take a look.

File: libcxx/test/libcxx/utilities/expected/expected.expected/transform_error.mandates.verify.cpp
Comment (philnik777) [inline libcxx/test/libcxx/utilities/expected/expected.expected/transform_error.mandates.verify.cpp:19]:
```suggestion
// TODO(LLVM-22): Remove '0-1' from 'expected-error-re@*:* 0-1 {{union member {{.*}} has reference type {{.*}}}}'
```
Same below.

**Review by philnik777 — APPROVED** [review]
LGTM with nit addressed.

File: libcxx/test/libcxx/utilities/expected/expected.expected/transform_error.mandates.verify.cpp
Comment (zeroomega) [reply libcxx/test/libcxx/utilities/expected/expected.expected/transform_error.mandates.verify.cpp]: Done.

**Comment by zeroomega** [comment]
Emmm, the upstream libcxx bot for msvc has an error that I didn't see when I local test it:

```
FAIL: llvm-libc++-shared-clangcl.cfg.in :: libcxx/utilities/expected/expected.void/transform_error.mandates.verify.cpp (1731 of 9739)
******************** TEST 'llvm-libc++-shared-clangcl.cfg.in :: libcxx/utilities/expected/expected.void/transform_error.mandates.verify.cpp' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# COMPILED WITH
'C:/Program Files/LLVM/bin/clang-cl.exe' D:\a\llvm-project\llvm-project\libcxx\test\libcxx\utilities\expected\expected.void\transform_error.mandates.verify.cpp --driver-mode=g++ --target=x86_64-pc-windows-msvc -fms-runtime-lib=dll -nostdinc++ -I D:/a/llvm-project/llvm-project/build/clang-cl-dll/include/c++/v1 -I D:/a/llvm-project/llvm-project/build/clang-cl-dll/include/c++/v1 -I D:/a/llvm-project/llvm-project/libcxx/test/support -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_STDIO_ISO_WIDE_SPECIFIERS -DNOMINMAX -std=c++26 -Werror -Wall -Wctad-maybe-unsupported -Wextra -Wshadow -Wundef -Wunused-template -Wno-unused-command-line-argument -Wno-attributes -Wno-pessimizing-move -Wno-noexcept-type -Wno-atomic-alignment -Wno-reserved-module-identifier -Wdeprecated-copy -Wdeprecated-copy-dtor -Wno-user-defined-literals -Wno-tautological-compare -Wsign-compare -Wunused-variable -Wunused-parameter -Wunreachable-code -Wno-unused-local-typedef -Wno-local-type-template-args -Wno-c++11-extensions -Wno-unknown-pragmas -Wno-pass-failed -Wno-mismatched-new-delete -Wno-redundant-move -Wno-self-move -D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_NONE -Werror=thread-safety -Wuser-defined-warnings -DTEST_WINDOWS_DLL  -fsyntax-only -Wno-error -Xclang -verify -Xclang -verify-ignore-unexpected=note -ferror-limit=0
# executed command: 'C:/Program Files/LLVM/bin/clang-cl.exe' 'D:\a\llvm-project\llvm-project\libcxx\test\libcxx\utilities\expected\expected.void\transform_error.mandates.verify.cpp' --driver-mode=g++ --target=x86_64-pc-windows-msvc -fms-runtime-lib=dll -nostdinc++ -I D:/a/llvm-project/llvm-project/build/clang-cl-dll/include/c++/v1 -I D:/a/llvm-project/llvm-project/build/clang-cl-dll/include/c++/v1 -I D:/a/llvm-project/llvm-project/libcxx/test/support -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_STDIO_ISO_WIDE_SPECIFIERS -DNOMINMAX -std=c++26 -Werror -Wall -Wctad-maybe-unsupported -Wextra -Wshadow -Wundef -Wunused-template -Wno-unused-command-line-argument -Wno-attributes -Wno-pessimizing-move -Wno-noexcept-type -Wno-atomic-alignment -Wno-reserved-module-identifier -Wdeprecated-copy -Wdeprecated-copy-dtor -Wno-user-defined-literals -Wno-tautological-compare -Wsign-compare -Wunused-variable -Wunused-parameter -Wunreachable-code -Wno-unused-local-typedef -Wno-local-type-template-args -Wno-c++11-extensions -Wno-unknown-pragmas -Wno-pass-failed -Wno-mismatched-new-delete -Wno-redundant-move -Wno-self-move -D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_NONE -Werror=thread-safety -Wuser-defined-warnings -DTEST_WINDOWS_DLL -fsyntax-only -Wno-error -Xclang -verify -Xclang -verify-ignore-unexpected=note -ferror-limit=0
# .---command stderr------------
# | error: 'expected-error' diagnostics seen but not expected: 
# |   File D:/a/llvm-project/llvm-project/build/clang-cl-dll/include/c++/v1\__expected/expected.h Line 107: call to deleted constructor of 'std::__expected_void_base<int &>::__repr'
# | 1 error generated.
# `-----------------------------
# error: command failed with exit status: 1
```

Looking into it.
We might have to update the test to add an expected-error string on this.

**Review by ldionne — APPROVED** [review]

**Comment by ldionne** [comment]
Do we need https://github.com/llvm/llvm-project/issues/104885 since we have a TODO(LLVM 22) in the code?

**Comment by ilovepi** [comment]
I asked for the bug so it wouldn’t get lost/forgotten, but that was for you (libc++ maintainers) , so if you’re going to track the cleanup differently, you’re free to close it.

**Comment by ldionne** [comment]
@zeroomega I'm fine either way. Can you link the bug in your comments? Another option would have been to add the `LLVM 22` milestone to the bug, but it doesn't exist yet. As-is, I fear that this bug would simply end up dangling. But if we reference it from the comment, I think that's fine.

**Comment by zeroomega** [comment]
> @zeroomega I'm fine either way. Can you link the bug in your comments? Another option would have been to add the `LLVM 22` milestone to the bug, but it doesn't exist yet. As-is, I fear that this bug would simply end up dangling. But if we reference it from the comment, I think that's fine.

Do you mean to link the bug in the comment section of `transform_error.mandates.verify.cpp` file or link the bug in the commit message?

**Comment by ldionne** [comment]
> > @zeroomega I'm fine either way. Can you link the bug in your comments? Another option would have been to add the `LLVM 22` milestone to the bug, but it doesn't exist yet. As-is, I fear that this bug would simply end up dangling. But if we reference it from the comment, I think that's fine.
> 
> Do you mean to link the bug in the comment section of `transform_error.mandates.verify.cpp` file or link the bug in the commit message?

In the comment in the code.

**Review by ldionne — APPROVED** [review]

**Pull request merged**
